flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -538,51 +542,61 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
-
-							Preconditions.checkState(
-								checkpoint != null || throwable != null,
-								"Either the pending checkpoint needs to be created or an error must have been occurred.");
-
-							if (throwable != null) {
-								// the initialization might not be finished yet
-								if (checkpoint == null) {
-									onTriggerFailure(request, throwable);
-								} else {
-									onTriggerFailure(checkpoint, throwable);
-								}
+			FutureUtils.waitForAll(asList(masterStatesComplete, coordinatorCheckpointsComplete))
+				.handleAsync(
+					(ignored, throwable) -> {
+						final PendingCheckpoint checkpoint =
+							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
+
+						Preconditions.checkState(
+							checkpoint != null || throwable != null,
+							"Either the pending checkpoint needs to be created or an error must have been occurred.");
+
+						if (throwable != null) {
+							// the initialization might not be finished yet
+							if (checkpoint == null) {
+								onTriggerFailure(request, throwable);
 							} else {
-								if (checkpoint.isDiscarded()) {
-									onTriggerFailure(
-										checkpoint,
-										new CheckpointException(
-											CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
-											checkpoint.getFailureCause()));
-								} else {
-									// no exception, no discarding, everything is OK
-									final long checkpointId = checkpoint.getCheckpointId();
-									snapshotTaskState(
-										timestamp,
-										checkpointId,
-										checkpoint.getCheckpointStorageLocation(),
-										request.props,
-										executions,
-										request.advanceToEndOfTime);
-
-									coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
-
-									onTriggerSuccess();
-								}
+								onTriggerFailure(checkpoint, throwable);
 							}
+						} else {
+							if (checkpoint.isDiscarded()) {
+								onTriggerFailure(
+									checkpoint,
+									new CheckpointException(
+										CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
+										checkpoint.getFailureCause()));
+							} else {
+								// no exception, no discarding, everything is OK
+								final long checkpointId = checkpoint.getCheckpointId();
+								snapshotTaskState(
+									timestamp,
+									checkpointId,
+									checkpoint.getCheckpointStorageLocation(),
+									request.props,
+									executions,
+									request.advanceToEndOfTime);
+
+								coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
+
+								onTriggerSuccess();
+							}
+						}
 
-							return null;
-						},
-						timer));
+						return null;
+					},
+					timer)
+				.whenComplete((unused, error) -> {
+					if (error != null) {
+						if (!isShutdown()) {
+							failureManager.handleJobLevelCheckpointException(new CheckpointException(EXCEPTION, error), Optional.empty());
+						} else if (error instanceof RejectedExecutionException) {
```

Comment:
Should we check the `RejectedExecutionException` here? It's kind of case by case. Do you think `log.warn` for every exception is better? In the future we could shut down timer and `CheckpointCoordinator` in a more elegant way to avoid this warning log.

Reply:
I think we can say that we tolerate exception after the shut down because we have closed the `timer` and can no longer guarantee that things keep working.

Reply:
Yes, I think we **should** check exception type, because `RejectedExecutionException` during shutdown is expected (though we should fix it with `RejectedExecutionHandler` in the future).

Other exceptions, even during the shutdown are not expected. They also might have occurred before the shutdown.

Reply:
But once the `CheckpointCoordinator` is shut down, these exceptions should no longer matter, at least not in terms of correctness of the job. I don't really understand why this was not a problem before and now we want to fine grained filter out exceptions, this does not seem like a straight forward solution to me.

Reply:
> But once the CheckpointCoordinator is shut down, these exceptions should no longer matter, at least not in terms of correctness of the job. 

So we are talkiing about the case `isShutdown() == true` branch. 
I agree about job correctness and in this case exceptions are only logged.
I think this can be helpful during troubleshooting or debugging.

> I don't really understand why this was not a problem before and now we want to fine grained filter out exceptions, this does not seem like a straight forward solution to me.

We didn't have this code in previous releases. `CheckpointCoordinator` was heavily refactored this release (than partially reverted). During the development, we did have problem recently and had to guess that there is an NPE.

Reply:
Yeah but the NPE occurred while the `CheckpointCoordinator` was not shut down. Hence it would be properly logged. Can you describe me a situation where it would matter to log an exception after the shut down?

Reply:
That NPE was a programming error and it could happen before the shutdown, for example in `onTriggerFailure`.